ruby/phase-1.md

```diff
@@ -0,0 +1,398 @@
+```
+
+Despite the differences both clients offer exactly the same interface (the singleton client is an
+instance of a regular client stored on Temporal module for convenience). The interface will look
+something like this:
+
+```ruby
+class Client
+  def initialize(config)
+
+  def start_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options overrides
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> WorkflowHandle
+
+  def workflow_handle(namespace, workflow_id, run_id = nil) -> WorkflowHandle
+end
+```
+
+Notes:
+
+- This is not a complete definition of the client
+- Both named and keyword arguments are supported, except for the reserved `options:`
+
+The combination of the singleton client and implicit execution attributes allows us to optionally
+support direct workflow invocations (removing all SDK traces out of the way):
+
+```ruby
+MyWorkflow.start(arg_1, arg_2, keyword_arg: arg_3)
+```
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a namespace/workflow_id/run_id combination
+using the `Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class WorkflowHandle
+  def initialize(connection, namespace, workflow_id, run_id = nil)
+
+  def result(await: false, timeout: nil) -> Any
+
+  def describe -> WorkflowExecutionInfo
+
+  def cancel -> void
+
+  def query(name, *args) -> Any
+
+  def signal(name, *args) -> void
+
+  def terminate(reason, details: nil) -> void
+end
+
+```
+
+
+## Worker
+
+The worker interface allows SDK users to register any number of both workflows and activities. These
+will then be polled for on the specified namespaces & task queues and executed. The worker can poll
+on multiple task queues at the same time. These will be determined by the registered activities and
+workflow. If there are any overlaps we'll only create one poller for each unique task queue.
+
+```ruby
+config Temporal::Configuration.new(url: 'localhost:7233')
+worker = Temporal::Worker.new(config)
+
+worker.register_workflow(NewWorkflow)
+worker.register_workflow(NewWorkflow, task_queue: 'tq-3')
+worker.register_activity(AnotherActivity)
+worker.register_activity(AnotherActivity, namespace: 'ns-3')
+
+# Calling Worker#run will block the main thread until receiving SIGTERM or SIGINT
+worker.run
+```
+
+Notes:
+
+- The worker will poll for workflows on `ns-1`/`tq-1` and `ns-1`/`tq-3`
+- And activities on `ns-2`/`tq-2` and `ns-3`/`tq-2`
+- The workflows/activities can of course be defined in any number of different workers
+- The worker is intended to take over the process
+
+### Concurrency model
+
+Ruby can be a tricky language for running concurrent applications. It supports kernel threads,
+however it has a Global Interpreter Lock (GIL) in place which prevents any two threads to be
```

Comment:
This is specific to MRI. TruffleRuby has no GIL and JRuby doesn't either. Both are potentially significant targets for heavy workloads.

Reply:
Good point, I'll add a note. I remember there were some issues with native extensions and JRuby. I wonder if that's still true and we won't be able to use it because of the Rust Core